pkg/skaffold/yamltags/paths.go

```diff
@@ -0,0 +1,128 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package yamltags
+
+import (
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+)
+
+// SetAbsFilePaths recursively sets all fields marked with the yamltag `filepath` to absolute paths
+func SetAbsFilePaths(s interface{}, base string) error {
+	errs := setAbsFilePaths(s, base)
+	if len(errs) == 0 {
+		return nil
+	}
+	var messages []string
+	for _, err := range errs {
+		messages = append(messages, err.Error())
+	}
+	return fmt.Errorf(strings.Join(messages, " | "))
+}
+
+func setAbsFilePaths(config interface{}, base string) []error {
+	if config == nil {
+		return nil
+	}
+	parentStruct := reflect.Indirect(reflect.ValueOf(config))
+
+	switch parentStruct.Kind() {
+	case reflect.Struct:
+		t := parentStruct.Type()
+		var errs []error
+		for i := 0; i < t.NumField(); i++ {
+			f := t.Field(i)
+			v := parentStruct.Field(i)
+			if !v.CanInterface() {
+				return errs
+			}
+			if filepathTagExists(f) {
+				switch v.Interface().(type) {
+				case string:
+					path := v.String()
+					if path == "" {
+						return errs
+					}
+					if filepath.IsAbs(path) {
+						return errs
+					}
+					v.SetString(filepath.Join(base, path))
+					logrus.Tracef("setting absolute path for config field %q", f.Name)
+				case []string:
+					for i := 0; i < v.Len(); i++ {
+						elem := v.Index(i)
+						path := elem.String()
+						if path == "" {
+							continue
+						}
+						if filepath.IsAbs(path) {
+							continue
+						}
+						elem.SetString(filepath.Join(base, path))
+						logrus.Tracef("setting absolute paths for config field %q index %d", f.Name, i)
+					}
+				default:
+					return []error{fmt.Errorf("yaml tag `filepath` needs struct field %q to be string or string slice", f.Name)}
+				}
+				return errs
+			}
+
+			if v.Kind() != reflect.Ptr {
+				v = v.Addr()
+			}
+			if elemErrs := setAbsFilePaths(v.Interface(), base); elemErrs != nil {
+				errs = append(errs, elemErrs...)
+			}
+		}
+		return errs
+	case reflect.Slice:
+		var errs []error
+		for i := 0; i < parentStruct.Len(); i++ {
+			elem := parentStruct.Index(i)
+			if elem.Kind() != reflect.Ptr {
+				elem = elem.Addr()
+			}
+			if !elem.CanInterface() {
+				continue
+			}
+			if elemErrs := setAbsFilePaths(elem.Interface(), base); elemErrs != nil {
+				errs = append(errs, elemErrs...)
+			}
+		}
+		return errs
+	default:
+		return nil
+	}
+}
+
+func filepathTagExists(f reflect.StructField) bool {
+	yamltags, ok := f.Tag.Lookup("yamltags")
```

Comment:
`yamltags` just doesn't seem to be appropriate.  Is there some prior art that we could leverage here, like the [Go validator's `dir` and `file` types?](https://github.com/go-playground/validator#other)?

Reply:
so do you mean like this?
```go
type Artifact struct {
	// Workspace is the directory containing the artifact's sources.
	// Defaults to `.`.
	Workspace string `yaml:"context,omitempty" validate:"dir"`
        ...
}
```

We are already using `yamltags` as the key for custom validation tags. Although I don't mind making this one different.

Reply:
maybe `convert: "dir"`? I think validate isn't correct either since we aren't validating the actual directory exists.

Reply:
I think that the [`validator`](https://godoc.org/gopkg.in/go-playground/validator.v9) package is validating the dir or file exists on the filesystem, which isn't what we're trying to do.

Reply:
Although we are bashing these values to make them reference to files and directories on disk, these values may be relative to places other than the current directory (e.g., within an artifact workspace).